### ISSUE TITLE
Installers allow you to attempt two installations at once, resulting in the loss of all settings.

### ISSUE THREAD

**Issue by stawberri** [OpenCollar/opencollar#970]
# Steps to reproduce
1. Rez a collar.
2. Configure some setting on your collar. For example, `<prefix> safeword EXAMPLE`.
3. Rez two copies of the OpenCollar Patch.
4. Start one of the updates.
5. While the other update has reached at least 25% but hasn't completed yet, start the other.
# Results
- The first patch will become stuck.
- Your collar will lose all of its settings. Check this by typing `<prefix> safeword` and see that it has reset.
# Real world scenario:

In my steps, I suggested using two patches because they're easy to obtain and take forever to install. When I actually discovered this bug, I was trying to update and install an app at the same time.
# My best guess of the cause
## Patch getting stuck

The first installer's `oc_update_shim` gets replaced by the second's.
## Settings loss

`oc_update_shim` seems to save current settings, update scripts, and then restore the settings. The installer that runs second may be reading the settings when the scripts no longer have them, and then restoring those empty settings onto the collar.
# Potential solution

At some point in the beginning of the installation process, check if there is already an installation happening, and refuse to complete the installation.

**Comment by GarvinTwine** [comment]
We already try to avoid this scenario when you start the update/installation from the collar by the "Update" button or command.
But right, if someone does it like you did, things don't go well.
At least for updates this could be solved by simply keeping the oc_sys script (which gets alwasy as first installed) not running until the update is done, so no 2nd installation can get started as this script provides the pin to enable to load scripts.
But this doesnt solve it if someone rezzes 5 different Apps and an update and just clicks them quickly one after the other... Here we can add a scan (same as the sys script does when the Update command is issued) to see if there are other installers/updaters around and if... we simply do nothing or even delete ourselves.
I just fear we might at some point run into one issue we won't be able to fix:

> A common mistake that people make when trying to design something completely foolproof is to underestimate the ingenuity of complete fools.

Douglas Adams

**Comment by stawberri** [comment]
In that case, maybe we could try something simpler, like adding a message to the installation prompt warning people not to do two installations at once? There's still room for error, but if a "WARNING" message was there I (probably) wouldn't have made this mistake.

**Comment by WendyStarfall** [comment]
I'll go out on a limb here and tell you all that I'm less than convinced of this fix/change.

Please keep in mind that updating a complex device such as the OpenCollar is a major undertaking for most users on the Second Life® platform. They are used to receive software updates automagically and if they have to click somewhere, confirm something, many users get quickly annoyed and interpret the update as something negative. Unfortunately we can't automatically deploy updates, this will always require user input.

Now, I believe that we managed to make this as painless as somehow possible on this platform. The input is virtually just taking the thing, touching it and clicking "Yes". However, teasing this routine that so many are already used to, like by introducing two notifies 🙊 when the installer is rezzed, is a recipe for disaster.

This gets especially mind-boggling when the installer completes, informs the user and naturally resets itself for the next installation, this way bopping the completion message and changelog links out of the user's view by notifying them "that they can now start with the installation". That's just. You know... no.

~~I will ship 6.3.3 without notifies.~~ Probably @GarvinTwine's original idea on how to tackle this would have been the better way to solve it. I'm just concerned that we add a whole lot of complexity to address an incident that happened in the last 5 years and 3 months exactly once.

Some more ideas:
- Move the "Please stand by.." message to the hovertext, i.e. "Intializing..."
- Avoid chatter like "Ready!" / "Touch now to start!" / "Proceed with blahblahblah..."
- Only chatter and inform the user if there is indeed a problem with multiple installers
- Add an integer to turn this whole circus off so I can deploy without having to wait for scans

tl;dr - avoid weird complexity; don't irritate the user; don't mess with the maintainer's work-flow (deploy)

**Comment by WendyStarfall** [comment]
I've branched off to [alt.installer](https://github.com/VirtualDisgrace/opencollar/tree/alt.installer) at https://github.com/VirtualDisgrace/opencollar/commit/549fbda0fbc638190c2b2043a01abad4fb430fe8 in our history to keep this variation of `oc_installer_sys` conveniently available for now (and for all those who might not be so familiar and comfy with git yet).

I will only keep the "Say function" which I believe makes sense because of our super long installer names. The rest of this will be addressed in the readme card for the installer and on the website with a note that multiple installations shouldn't be done at the same time.
